src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -597,35 +598,35 @@ internal MemberAccessor MemberAccessorStrategy
         }
 
         /// <summary>
-        /// Whether the options instance has been primed for reflection-based serialization.
+        /// JsonTypeInfo resolver used for serialization.
         /// </summary>
-        internal bool IsInitializedForReflectionSerializer;
+        internal IJsonTypeInfoResolver? _typeInfoResolver;
```

Comment:
Shouldn't the `IJsonTypeInfoResolver` field be replacing the existing `_serializerContext` field?

https://github.com/dotnet/runtime/blob/86bbc2bafde47db66f57f980ef5dcf5f09db85b7/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L38

I would recommend unifying the two and preserving the sentinel so that fallback to the default resolver singleton is preserved.

Reply:
I'll need to leave it to preserve current behavior (i.e. making options immutable once context set) and making it easier to preserve current fallback logic but I can scope the usage to just that and use is through `_typeInfoResolver`. Let's do the breaking changes separately

Reply:
I agree we should do breaking changes separately, but I think the first step in introducing `IJsonTypeInfo` is implementing it where applicable and make refactorings as appropriate. Merging the two fields doesn't mean we're introducing any breaking changes.

Reply:
I've kept it in this iteration but only use it to preserve current behavior (more like sentinel)

Reply:
I would recommend removing the `JsonSerializerContext` field and wire everything around using `IJsonTypeInfo` for metadata resolution. No point refactoring this code now if we're going to change it again when it does get removed.